### PR TITLE
Ts rewrite

### PR DESCRIPTION
**Pull request by mo-gr**
I needed a `d.ts` file for perron. Somehow this ended up turning into a TS rewrite... This is probably not a fun PR to review...

In theory, it should be 100% backwards compatible and still uses old tests in js. Once, we agree this is a stable rewrite, I will move the tests to ts as well. But doing both in one go felt very risky.

However, I also went to standard.js for the tests and they therefore show diffs as well. Maybe that wasn't too smart...

Maybe it is best, to try this version in projects, that actually use perron and see if they still work. (I did this for a bigger, internal one and it still worked as expected.)

### CODE REVIEW

File: lib/client.ts [hunk 1]
```diff
@@ -0,0 +1,361 @@
+import { ServiceClientRequestOptions, ServiceClientResponse, request } from './request'
+import * as url from 'url'
+
+const retry = require('retry')
```
Comment (grassator) [inline lib/client.ts]: Would be nice to have a note why these are `require` statements and not `import`

File: lib/client.ts [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+export interface ServiceClientRequestFilter {
+  /**
+   * This callback is called before the requests is done.
+   * You can short-circuit the request by both returning
+   * a ServiceClient.Response object which is helpful for
+   * implementing caching or mocking. You could also
+   * fail the request by throwing an Error.
+   * @throws {Error}
+   */
+  request?: (requestOptions: ServiceClientRequestOptions) => Promise<ServiceClientResponse| ServiceClientRequestOptions>
```
Comment (grassator) [inline lib/client.ts]: Formatting, a space before `|`. Maybe we need to add this rule to TSLint

File: lib/client.ts [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    }
+  }
+  return response;
+}
+
+/**
+ * Wrapper that makes sure that all error coming out
+ * of ServiceClients are actual ServiceClientError
+ */
+const wrapFailedError = (type: string, error: Error | ServiceClientError, responseThunk?: Function) => {
```
Comment (grassator) [inline lib/client.ts]: It doesn't add too much, but I would change type from `Function` to `() => any`, which tells that we don't expect any arguments

File: lib/client.ts (hunk 1 above)
Comment (mo-gr) [reply lib/client.ts]: would be even nicer to have them proper imports -_-'

I'll add a note

File: lib/client.ts (hunk 2 above)
Comment (mo-gr) [reply lib/client.ts]: standard-ts?

File: lib/client.ts (hunk 3 above)
Comment (mo-gr) [reply lib/client.ts]: will do

**Comment by grassator** [comment]
👍

**Comment by grassator** [comment]
@shuhei I would really appreciate a second pair of eyes on this

**Comment by shuhei** [comment]
👍

**Pull request merged**